### ISSUE TITLE
[BUG] When we ssh to VM, the IP Address disappear on the virtual machine page

### ISSUE THREAD

**Issue by TachunLin** [yasker/vm#3990]
**Describe the bug**

When we ssh to VM, the allocated IP Address from `vlan1` network would disappear from Virtual machine

https://github.com/harvester/harvester/assets/29251855/6bedaf28-c8e7-40ad-8e1f-5c394037cb65

![image](https://github.com/harvester/harvester/assets/29251855/dab27d9f-dd73-4d49-954e-1dc65d411981)

**To Reproduce**
1. Upload a opensuse 15.3 leap no cloud image 
2. Create a `vlan1` vlan network on `mgmt` interface
3. Create a volume 
4. Create a VM with opensue 15.3 leap image and use `vlan1` network
5. Wait the VM started and display IP address on VM page 
6. ssh to the VM 
7. Check the IP address of VM on VM page and edit config

**Expected behavior**

After ssh to VM, the IP Address should not disappear on VM page

**Support bundle**

[supportbundle_fc8c5ccb-0eab-4b54-902c-fd4faad9b9fe_2023-05-29T08-54-45Z.zip](https://github.com/harvester/harvester/files/11590233/supportbundle_fc8c5ccb-0eab-4b54-902c-fd4faad9b9fe_2023-05-29T08-54-45Z.zip)

**Environment**
 - Harvester ISO version: `master-35717460-head` (05/29)
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Single nodes on local kvm machine

**Additional context**
1. After several minutes the IP Address will come back
2. Currently exist on opensuse leap image 

* Used image url: 
http://download.opensuse.org/repositories/Cloud:/Images:/Leap_15.3/images/openSUSE-Leap-15.3.x86_64-NoCloud.qcow2

**Comment by rrajendran17** [comment]
@TachunLin I tried this on v1.3.1, and I do not see the issue. (Tried few iterations of same steps)
Can you please confirm if this might be fixed later after you saw the issue ? I do not see much clue from the support bundle collected as well.We might need the vm console logs during the issue state.

**Comment by bk201** [comment]
@rrajendran17 we have another similar case: https://github.com/harvester/harvester/issues/5251#issuecomment-2273952105
In this report, the cause seems to be the qemu-agent version.

@bk201 Thanks for the reference to other bug

**Comment by rrajendran17** [comment]
@TachunLin @bk201 I am able to reproduce this issue locally consistently by creating multiple vms at the same time(10 vms).
I see the issue is seen in opensuse-leap-15.3 as well as opensuse-leap-15-6 images(this uses latest qemu-ga 8.2.5 version).
So in this bug the issue is not related to qemu-guest-agent.

I see the vm ips in the GUI flaps for once and then stays intact and did not disappear after that.
So I feel this could be related some interface reconfigured during init.I am debugging this further to find the root cause the issue.

**Comment by rrajendran17** [comment]
Looks like the issue is due to the sequence of ips which kubevirt sees.
Kubevirt sees the ip from domain(pod ip) first which is given to the harvester UI

interfaces:
  - infoSource: domain
    ipAddress: 10.52.0.183
    ipAddresses:
    - 10.52.0.183
    mac: 1a:8e:0d:da:18:d2

and later there is interface information from guest agent without ip address (while ip address was tried for dhcp)
which is updated to harvester UI (ip disappears from UI)

- infoSource: domain, guest-agent
    interfaceName: eth0
    mac: 1a:8e:0d:da:18:d2

and later guest-agent/domain updates the ip address on the interface to kubevirt and again ip address is updated in harvester UI(ip address appears in UI)

interfaces:
  - infoSource: domain, guest-agent
    interfaceName: eth0
    ipAddress: 10.52.0.183
    ipAddresses:
    - 10.52.0.183
    mac: 1a:8e:0d:da:18:d2

Need to see how harvester UI fetches this info and updates UI, and see if address from just domain could be skipped.
The address should be updated in UI only when interface info(interfaceName) and by guest-agent is available.

**Comment by rrajendran17** [comment]
Note: Above issue is seen only with mgmt interface and I am not able to reproduce with bridge interface.I tried with both ubuntu and opensuse-leap(15.3,15.6).
@TachunLin can you please confirm if this issue is still seen on vlan vm network ips ?

**Comment by vwbusguy** [comment]
> Above issue is seen only with mgmt interface and I am not able to reproduce with bridge interface.

If it only impacts the management interface then this is a different issue than https://github.com/harvester/harvester/issues/5251#issuecomment-2273952105 as that issue impacts bridge interface IPs as well.  (For context, I work in the same org as ParadoxGuitarist.)

**Comment by rrajendran17** [comment]
@vwbusguy  yes this issue is completely different from https://github.com/harvester/harvester/issues/5251 not only from bridge interface ip perspective but also this issue happens only during start up (ip disappears  only once during start up and then comes back) and stays intact in UI after that.

**Comment by albinsun** [comment]
I hit another case that opensuse VM IP will disappear intermittently w/o any operations.
And the frequency of occurrence seems decreases over time (more stable over time)

Note that the root cause may relates to guest OS cuz we do not observe this on Ubuntu.

[supportbundle_restore_new_ip_2024-08-22T07-11-37Z.zip](https://github.com/user-attachments/files/16726518/supportbundle_restore_new_ip_2024-08-22T07-11-37Z.zip)

https://github.com/user-attachments/assets/951777e7-57d5-47ce-8eb4-a592b05a925d

**Comment by rrajendran17** [comment]
With additional debugs enabled at qemu-ga and libvirt, it shows the interface statuses are retrieved and updated periodically every 120 seconds (2 minutes) by virt-launcher as expected, but after every few seconds after this updation, the harvester node UI loses the ip or the ip disappears in the UI.

2024-08-23 18:18:01.790+0000: 29: debug : qemuAgentCommandFull:1039 : Send command '{"execute":"guest-network-get-interfaces"}' for write, seconds = -1
2024-08-23 18:18:01.791+0000: 82: debug : qemuAgentIOProcessLine:214 : Line [{"return": [{"name": "lo", "ip-addresses": [{"ip-address-type": "ipv4", "ip-address": "127.0.0.1", "prefix": 8}, {"ip-address-type": "ipv6", "ip-address": "::1", "prefix": 128}], "statistics": {"tx-packets": 30, "tx-errs": 0, "rx-bytes": 2432, "rx-dropped": 0, "rx-packets": 30, "rx-errs": 0, "tx-bytes": 2432, "tx-dropped": 0}, "hardware-address": "00:00:00:00:00:00"}, {"name": "eth0", "ip-addresses": [{"ip-address-type": "ipv4", "ip-address": "172.19.99.108", "prefix": 21}, {"ip-address-type": "ipv6", "ip-address": "fe80::3845:5ff:feb8:838d", "prefix": 64}], "statistics": {"tx-packets": 15287, "tx-errs": 0, "rx-bytes": 53774336, "rx-dropped": 55065, "rx-packets": 337936, "rx-errs": 0, "tx-bytes": 2845439, "tx-dropped": 0}, "hardware-address": "3a:45:05:b8:83:8d"}]}]
2024-08-23 18:18:01.791+0000: 82: debug : qemuAgentIOProcessData:294 : Total used 776 bytes out of 776 available in buffer

2024-08-23 18:20:01.790+0000: 31: debug : qemuAgentCommandFull:1039 : Send command '{"execute":"guest-network-get-interfaces"}' for write, seconds = -1
2024-08-23 18:20:01.791+0000: 82: debug : qemuAgentIOProcessLine:214 : Line [{"return": [{"name": "lo", "ip-addresses": [{"ip-address-type": "ipv4", "ip-address": "127.0.0.1", "prefix": 8}, {"ip-address-type": "ipv6", "ip-address": "::1", "prefix": 128}], "statistics": {"tx-packets": 30, "tx-errs": 0, "rx-bytes": 2432, "rx-dropped": 0, "rx-packets": 30, "rx-errs": 0, "tx-bytes": 2432, "tx-dropped": 0}, "hardware-address": "00:00:00:00:00:00"}, {"name": "eth0", "ip-addresses": [{"ip-address-type": "ipv4", "ip-address": "172.19.99.108", "prefix": 21}, {"ip-address-type": "ipv6", "ip-address": "fe80::3845:5ff:feb8:838d", "prefix": 64}], "statistics": {"tx-packets": 15291, "tx-errs": 0, "rx-bytes": 53817315, "rx-dropped": 55165, "rx-packets": 338497, "rx-errs": 0, "tx-bytes": 2845703, "tx-dropped": 0}, "hardware-address": "3a:45:05:b8:83:8d"}]}]
2024-08-23 18:20:01.791+0000: 82: debug : qemuAgentIOProcessData:294 : Total used 776 bytes out of 776 available in buffer

When manually executed a command "kubectl get vmi <vmi-name> -w -o yaml" to get vmi status, I do not see any explicit qemuAgentCommand in qemu-ga, libvirt logs to fetch the interface statuses information.Need to add more debug logs to virt-launcher pod to understand the issue.

**Comment by w13915984028** [comment]
FYI: The qemu agent retured yaml is parsed via

https://github.com/kubevirt/kubevirt/blob/a8f8c2bdc2e2502425fa1bdb5db449ef8cc48d8f/pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go#L264

**Comment by rrajendran17** [comment]
With additional debugs enabled in virt-launcher and virt-handler pod, I could see that "guest-network-get-interfaces" is retrieved by virt-launcher polling agent every 2 minutes and the eventChannel is updated correctly for domain.Status.Interfaces.

https://github.com/kubevirt/kubevirt/blob/c70f865e30ab7c19da96c395e42623d7e07e9d8c/pkg/virt-launcher/notify-client/client.go#L438

eventCallback updates the domain.Status.Interfaces whenever there is update to interface (interfaceStatus not nil), but during the issue state, I see domain.Status.Interfaces becomes empty randomly though its not updated during the flow.
The same eventCallback could be called whenever there is an update to other qemu-ga data like fsinfo,os-info etc, but interfaceStatus will be updated only if interfaceStatus is not nil, but domain.Status.Interface becoming empty makes me believe that in the same flow domain.Status.Interface might have been wiped off during other updates to the domain struct, though I am not sure what could be causing it.

I believe that there could be a data corruption or concurrency issue which is causing domain.Status.Interfaces to become empty.

https://github.com/kubevirt/kubevirt/blob/c70f865e30ab7c19da96c395e42623d7e07e9d8c/pkg/virt-launcher/notify-client/client.go#L345

In the following logs, "copy intf" is domain.Status.Interfaces and "org" is InterfaceStatus coming from virt-launche poller notification.

During issue state:
logs:7796:2024-08-28T04:01:56.913108167Z {"component":"virt-launcher","level":"info","msg":"eventCallback interfaceStatus is nil","pos":"client.go:348","timestamp":"2024-08-28T04:01:56.913016Z"}
logs:7797:2024-08-28T04:01:56.913121567Z {"component":"virt-launcher","level":"info","msg":"eventCallback interfaceStatus copy intf [] org []","pos":"client.go:358","timestamp":"2024-08-28T04:01:56.913050Z"}

During non issue state:
ogs:185:2024-08-28T02:31:29.365918888Z {"component":"virt-launcher","level":"info","msg":"eventCallback interfaceStatus is nil","pos":"client.go:348","timestamp":"2024-08-28T02:31:29.365805Z"}
logs:186:2024-08-28T02:31:29.365932464Z {"component":"virt-launcher","level":"info","msg":"eventCallback interfaceStatus copy intf [{Mac:9a:26:5f:89:d0:c8 Ip:172.19.99.59 IPs:[172.19.99.59 fe80::9826:5fff:fe89:d0c8] InterfaceName:eth0}] org []","pos":"client.go:358","timestamp":"2024-08-28T02:31:29.365848Z"}

Summary:
This happens whenever there are frequent updates from qemu-ga on any of the guest commands polled from virt-launcher.This is seen with alma-linux  because there are frequent change in the information shared
from qemu-ga of alma-linux especially guest-get-fsinfo which  is not happening with other images like ubuntu etc.I believe we will see "ip disapperaring issue" on any vm running guest os which will have frequent updates from qemu-ga to the virt-launcher during the polling thread

**Comment by w13915984028** [comment]
Could you double check/confirm with following code change: thanks.

```
var interfaceStatuses []api.InterfaceStatus

		if interfaceStatus != nil {
			domain.Status.Interfaces = interfaceStatus
		}
```

->

```
		if interfaceStatus != nil && len(interfaceStatus) != 0 {
			domain.Status.Interfaces = interfaceStatus
		}
```

From the log, when the `interfaceStatus` is `[]`, it may also write to upstream ?

**Comment by rrajendran17** [comment]
Thanks @w13915984028 but I had added a check to confirm is interfaceStatus is nil everytime interfaceStatus is interfaceStatus[] and it prints nil and does not do any copy to domain.Status.Interface. So if the slice is empty, it will be nil, I believe explicit length check of the slice is not mandatory.
Though I can re confirm this once.

**Comment by w13915984028** [comment]
@rrajendran17 Thanks. Besides the callback, another known place is the vmi onchange, which will update the net status.

keyword is: updateVMIIfaceStatusWithGuestAgentData

**Comment by rrajendran17** [comment]
@w13915984028 yes my understanding is also same. virt-launcher eventupdate updates the domain.Status.Interfaces from data retrieved from qemu-ga periodically.And UpdateStatus in virt-handler updates the status into vmi object from domain.Status.Interfaces.
What we see in the output of kubectl get vmi <vmi-name> -o yaml (InfoSource, interfaces...) is from updateVMIIfaceStatusWithGuestAgentData by UpdateStatus from virt-handler

**Comment by rrajendran17** [comment]
The difference I see between alma-linux9-0507(issue) and alam-linux9-0805(non issue) is that, alma-linux9-0507 has more frequent updates of the filesysteminfo.("guest-get-fsinfo",GET_FILESYSTEM) and alam-linux9-0805 does not have change in fsinfo from qemu-ga.

From below code ,it is evident that, we send agentUpdate whenever there is any change from qemu-ga, so frequent changes to fs-info triggers an update along with sending domainInfo as well during which domain.Status.Interfaces becoming empty though it is not updated during the flow.(cause unknown)

https://github.com/kubevirt/kubevirt/blob/ad71d984f8a56740104da236c03979cd8a0217b7/pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go#L84

So the issue could be in kubevirt/virt-launcher along with trigger from underlying guest OS(whenever there is frequent update of fs-info during the agent polling)

The following code also looks incorrect to me as it sends update of domainInfo data in case of any change.Instead I think it should be only for (GET_OSINFO, GET_INTERFACES, GET_FSFREEZE_STATUS).If the aim is to only update domainInfo, then its not necessary to send update for other commands.

Existing code:
func (s *AsyncAgentStore) Store(key AgentCommand, value interface{}) {

	oldData, _ := s.store.Load(key)
	updated := (oldData == nil) || !equality.Semantic.DeepEqual(oldData, value)

	s.store.Store(key, value)

	if updated {
		domainInfo := api.DomainGuestInfo{}
		switch key {
		case GET_OSINFO, GET_INTERFACES, GET_FSFREEZE_STATUS:
			domainInfo.OSInfo = s.GetGuestOSInfo()
			domainInfo.Interfaces = s.GetInterfaceStatus()
			domainInfo.FSFreezeStatus = s.GetFSFreezeStatus()
		}

		s.AgentUpdated <- AgentUpdatedEvent{
			DomainInfo: domainInfo,
		}
	}
}

Instead is it supposed to do the following


func (s *AsyncAgentStore) Store(key AgentCommand, value interface{}) {

        oldData, _ := s.store.Load(key)

        s.store.Store(key, value)

        switch key {
        case GET_OSINFO, GET_INTERFACES, GET_FSFREEZE_STATUS:
                updated := (oldData == nil) || !equality.Semantic.DeepEqual(oldData, value)
                if updated {
                        domainInfo := api.DomainGuestInfo{}
                        domainInfo.OSInfo = s.GetGuestOSInfo()
                        domainInfo.Interfaces = s.GetInterfaceStatus()
                        domainInfo.FSFreezeStatus = s.GetFSFreezeStatus()
                        s.AgentUpdated <- AgentUpdatedEvent{
                                DomainInfo: domainInfo,
                        }
                }
        }
}

**Comment by albinsun** [comment]
> I hit another case that opensuse VM IP will disappear intermittently w/o any operations. And the frequency of occurrence seems decreases over time (more stable over time)
> 
> Note that the root cause may relates to guest OS cuz we do not observe this on Ubuntu.
> 
> [supportbundle_restore_new_ip_2024-08-22T07-11-37Z.zip](https://github.com/user-attachments/files/16726518/supportbundle_restore_new_ip_2024-08-22T07-11-37Z.zip)
> 
>  vokoscreenNG-2024-08-23_16-28-28-cut.mp4

Update a recent hit with SB.
As you can see, here we have **sles15sp6**, **sles15sp4**, **ubuntu24** and **rhel9**, and sles experience similiar behavior as this issue.
[supportbundle_UnstableIPShowing_2024-09-03T01-19-40Z.zip](https://github.com/user-attachments/files/16842314/supportbundle_UnstableIPShowing_2024-09-03T01-19-40Z.zip)

* ![image](https://github.com/user-attachments/assets/cd8a4591-abef-4092-b1b1-35da1f987f30)
* ![image](https://github.com/user-attachments/assets/c526bfc2-661f-4fe2-aaf0-5722368cbd50)

**Comment by rrajendran17** [comment]
@albinsun I have raised an issue with kubevirt community with analysis and following up with them on this.
https://github.com/kubevirt/kubevirt/issues/12698

**Comment by w13915984028** [comment]
The new SB attached by @albinsun provides solid evidence that the `vmi.status.interfaces is missing the related information
```
sles15sp4 vm:
    labels:
      harvesterhci.io/vmName: sles15sp4-20240903
      kubevirt.io/nodeName: harvester-node-1

    networks:
    - multus:
        networkName: default/ens5-vlan1
      name: default
      
    interfaces:
    - infoSource: domain, multus-status // IP is missing
      mac: a2:0d:db:c9:13:bd
      name: default
      queueCount: 1
      
the console log:
...
2024-09-03T01:06:25.408879877Z Welcome to SUSE Linux Enterprise Server 15 SP4  (x86_64) - Kernel 5.14.21-150400.22-default (ttyS0).
2024-09-03T01:06:25.408965276Z 
2024-09-03T01:06:25.408977214Z eth0: 192.168.121.143 fe80::a00d:dbff:fec9:13bd
...
      
//////////////////////////
ubuntu vm:

    hostname: ubuntu24-20240903
    
    networks:
    - multus:
        networkName: default/ens5-vlan1
      name: default
     
    interfaces:
    - infoSource: domain, guest-agent, multus-status
      interfaceName: enp1s0
      ipAddress: 192.168.121.137
      ipAddresses:
      - 192.168.121.137
      - fe80::309b:6aff:fee9:a014
      mac: 32:9b:6a:e9:a0:14
      name: default
      queueCount: 1     
```

**Comment by rrajendran17** [comment]
There is an update on this issue from Kubevirt community.
They are working on some change in interface status area.Will post their conclusion soon.

**Comment by Dung10012003** [comment]
![z6111838578943_2c9dfa035f836639f1044454ec52043b](https://github.com/user-attachments/assets/1d2b7e65-66f2-4248-8175-9d227b2c88b5)
![z6111838579019_90cf687f6b322e24fa51ad512c8e6beb](https://github.com/user-attachments/assets/c7918106-83a4-49f3-a345-b2540319c36d)
![z6111838579025_26cfb73ccb919423a28ec320990847df](https://github.com/user-attachments/assets/60e47131-fd09-4d1b-ae47-36403b1a7849)
I am taking 3 harvertes together. I've been using it for 2 weeks, it runs very stable. After I turned it off for 2 days, I got the error "server does not display cluster IP", please help. Thank you

**Comment by bk201** [comment]
@Dung10012003 I feel the issue you saw is not related to this one, do you mind creating another new issue? And please also provide more info: Like how many nodes there are in total, do rke2-server works on all nodes? Thanks.

**Comment by votdev** [comment]
Additional issues that are related to this problem:
- https://github.com/harvester/harvester/issues/5276
- https://github.com/harvester/harvester/issues/6644
- https://github.com/harvester/harvester/issues/7520

**Comment by rrajendran17** [comment]
@votdev  The following issues are not related to this.Though the symptoms are similar the cause is different.

https://github.com/harvester/harvester/issues/5276 - This is because secondary interfaces will be down and will not get ip address when not done manually or included in cloud init scripts.

https://github.com/harvester/harvester/issues/6644 - qemu-ga is not installed so ips are not updated in vmi.

https://github.com/harvester/harvester/issues/7520 -  This is another kubevirt bug fixed in latest releases where ipv4 will not be updated if ipv6 address is present (https://github.com/harvester/harvester/issues/6955)

The issue in this GH is due to a kubevirt bug where sometimes on some os, qemu-ga will not update the ip address to the vmi.
(related https://github.com/kubevirt/kubevirt/pull/13624)

**Comment by innobead** [comment]
@rrajendran17 the upstram PR was merged. Check if we can move this forward. 

> The issue in this GH is due to a kubevirt bug where sometimes on some os, qemu-ga will not update the ip address to the vmi. (related [kubevirt/kubevirt#13624](https://github.com/kubevirt/kubevirt/pull/13624))

Update: https://github.com/kubevirt/kubevirt/issues/12698#issuecomment-3243975884, that PR was just merged so that it won't be eligible for Harvester 1.7.0 but a later version instead.

**Comment by rrajendran17** [comment]
@innobead I will test once with kubevirt master branch to confirm it resolves the issue.And also I have requested kubevirt team about the release info in which the fix will be available.

**Comment by rrajendran17** [comment]
The upstream fix https://github.com/kubevirt/kubevirt/pull/13624 will be available in v1.7 kubevirt version

**Comment by JavierLagosChanclon** [comment]
Hey @rrajendran17. Have you tested the new KubeVirt version that contains the fix for this bug? In which Harvester version can we expect to have this KubeVirt version?

TIA!

**Comment by Vicente-Cheng** [comment]
> The upstream fix [kubevirt/kubevirt#13624](https://github.com/kubevirt/kubevirt/pull/13624) will be available in v1.7 kubevirt version

We will bump Kubevirt v1.7 with Harvester v1.8.0.
So, moved to `Icebox` first, after we bump Kubevirt v1.7.0, moving to the `Ready for Testing`

**Comment by rrajendran17** [comment]
Issue reproducible on Harvester version v1.7.1(kubevirt v1.6.3) with alma linux 9 0507 version and issue not seen on Harvester version v1.8.0 (kubevirt v1.7.0) as expected.Fixed by https://github.com/kubevirt/kubevirt/pull/13624. The IP in the UI stays intact. Moving to Ready to Testing.

<img width="1341" height="109" alt="Image" src="https://github.com/user-attachments/assets/8f51ac15-2b49-42e0-9385-fcd23494becd" />

alma linux version where the issue was reproducible -https://vault.almalinux.org/9.4/cloud/x86_64/images/AlmaLinux-9-GenericCloud-9.4-20240507.x86_64.qcow2

**Comment by lanfon72** [comment]
Closed since the issue could not be reproduced in `v1.8.0-dev-20260301` after several tests (kubevirt version: `v1.7.0`)

Test Information
----
* Environment: **MegaRAC SP-X 3 nodes**
* Harvester Version: **v1.8.0-dev-20260301**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Follow **To Reproduce** in https://github.com/harvester/harvester/issues/3990#issue-1730355430
